### PR TITLE
Add check for executor and observable compatibility in executor.py

### LINKED ISSUE #1140: Raise error if executor and observable are not compatible
Issue Description
-----------------

If an `executor` returns expectation values as `float` variables, Mitiq should always be used with `observable=None` (default),
since in this case the observable is implicitly defined and measured in the `executor`.

From a Discord discussion, it seems that no error is raised for incompatible settings.

Solution
-----------------

- Raise an error if the `observable` argument is not `None` and the executor doesn't return a `MeasurementLike`  result (bitstrings).

### PR DESCRIPTION
**Pull request by Misty-W**
Raise an error if the observable argument is not `None` and the executor doesn't return a "MeasurementLike" result (bitstrings).

Description
-----------

Fixes #1140

Checklist
---------

Check off the following once complete (or if not applicable) after opening the PR. The PR will be reviewed once this checklist is complete and all tests are passing.

- [x] I added unit tests for new code.
- [X] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [X] I used [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings for functions.
- [x] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.
- [X] Added myself / the copyright holder to the AUTHORS file

If some items remain, you can mark this a [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

License
-------

- [X] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Tips
----

- If the validation check fails:

    1. Run `make check-types` (from the root directory of the repository) and fix any [mypy](https://mypy.readthedocs.io/en/stable/) errors.

    2. Run `make check-style` and fix any [flake8](http://flake8.pycqa.org) errors.

    3. Run `make format` to format your code with the [black](https://black.readthedocs.io/en/stable/index.html) autoformatter.

  For more information, check the [Mitiq style guidelines](https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines).

- Write "Fixes #XYZ" in the description if this PR fixes Issue #XYZ.

### CODE REVIEW

**Comment by Misty-W** [comment]
Following tests are failing after adding observable and executor compatibility check. Some need to be modified before merge, others removed (and replaced in a follow-up issue).  

1. mitiq/cdr/tests/test_cdr.py::test_execute_with_cdr 
2. mitiq/cdr/tests/test_cdr.py::test_execute_with_cdr_using_clifford_circuit
3. mitiq/cdr/tests/test_cdr.py::test_execute_with_variable_noise_cdr 
4. mitiq/executor/tests/test_executor.py::test_executor_evaluate_density_matrix[executor_density_matrix_batched]
5. mitiq/observable/tests/test_observable.py::test_observable_expectation_one_circuit
6. mitiq/observable/tests/test_observable.py::test_observable_expectation_two_circuits
7. mitiq/observable/tests/test_observable.py::test_observable_expectation_supported_qubits
8. mitiq/raw/tests/test_raw.py::test_raw
9. mitiq/zne/tests/test_zne.py::test_with_observable_batched_factory
10. mitiq/zne/tests/test_zne.py::test_with_observable_adaptive_factory
11. mitiq/zne/tests/test_zne.py::test_with_observable_two_qubits
12. mitiq/pec/tests/test_pec.py::test_execute_with_pec_partial_representations
13. mitiq/pec/tests/test_pec.py::test_execute_with_pec_with_observable

**Comment by Misty-W** [comment]
Docs failing: 

1. observables.myst
2. cdr-3-options.myst
3. zne-3-options.myst
4. cdr-1-intro.myst
5. executors.myst

**Comment by Misty-W** [comment]
Failed in ASV benchmarks:
- benchmarks.track_pec

File: mitiq/executor/executor.py
Comment (andreamari) [inline mitiq/executor/executor.py:156]:
```suggestion
            and self._executor_return_type not in MeasurementResultLike
            and self._executor_return_type not in DensityMatrixLike
```

The observable is compatible with DensityMatrixLike return types too. This may fix most of the problems.

**Comment by andreamari** [comment]
Great @Misty-W! Tests are good now. The only thing which is now missing is a new test to check that the new error that you introduced in this PR is raised as expected. 
This should fix the code coverage too.

File: mitiq/executor/tests/test_executor.py
Comment (andreamari) [inline mitiq/executor/tests/test_executor.py:233]:
```suggestion
    with pytest.raises(ValueError, match="are not compatible"):
```

**Review by andreamari — APPROVED** [review]
Thanks @Misty-W! LGTM :+1:
I think @crazy4pi314 is testing the github pages action here. So I don't merge.

**Comment by Misty-W** [comment]
Closed #1140 via #1198 to work around GH action issue on this PR

**Pull request closed without merging**
